pkg/upgrade/base/upgrade_common.go

```diff
@@ -0,0 +1,857 @@
+	log logr.Logger,
+	k8sConfig *rest.Config,
+	scheme *runtime.Scheme,
+	eventRecorder record.EventRecorder) (*CommonUpgradeManagerImpl, error) {
+	k8sClient, err := client.New(k8sConfig, client.Options{Scheme: scheme})
+	if err != nil {
+		return &CommonUpgradeManagerImpl{}, fmt.Errorf("error creating k8s client: %v", err)
+	}
+
+	k8sInterface, err := kubernetes.NewForConfig(k8sConfig)
+	if err != nil {
+		return &CommonUpgradeManagerImpl{}, fmt.Errorf("error creating k8s interface: %v", err)
+	}
+
+	nodeUpgradeStateProvider := NewNodeUpgradeStateProvider(k8sClient, log, eventRecorder)
+	commonUpgrade := CommonUpgradeManagerImpl{
+		Log:                      log,
+		K8sClient:                k8sClient,
+		K8sInterface:             k8sInterface,
+		EventRecorder:            eventRecorder,
+		DrainManager:             NewDrainManager(k8sInterface, nodeUpgradeStateProvider, log, eventRecorder),
+		PodManager:               NewPodManager(k8sInterface, nodeUpgradeStateProvider, log, nil, eventRecorder),
+		CordonManager:            NewCordonManager(k8sInterface, log),
+		NodeUpgradeStateProvider: nodeUpgradeStateProvider,
+		ValidationManager:        NewValidationManager(k8sInterface, log, eventRecorder, nodeUpgradeStateProvider, ""),
+		SafeDriverLoadManager:    NewSafeDriverLoadManager(nodeUpgradeStateProvider, log),
+	}
+	return &commonUpgrade, nil
+}
+
+// WithPodDeletionEnabled provides an option to enable the optional 'pod-deletion' state and pass a custom
+// PodDeletionFilter to use
+func (m *CommonUpgradeManagerImpl) WithPodDeletionEnabled(filter PodDeletionFilter) CommonUpgradeStateManager {
+	if filter == nil {
+		m.Log.V(consts.LogLevelWarning).Info("Cannot enable PodDeletion state as PodDeletionFilter is nil")
+		return m
+	}
+	m.PodManager = NewPodManager(m.K8sInterface, m.NodeUpgradeStateProvider, m.Log, filter, m.EventRecorder)
+	m.podDeletionStateEnabled = true
+	return m
+}
+
+// WithValidationEnabled provides an option to enable the optional 'validation' state and pass a podSelector to specify
+// which pods are performing the validation
+func (m *CommonUpgradeManagerImpl) WithValidationEnabled(podSelector string) CommonUpgradeStateManager {
+	if podSelector == "" {
+		m.Log.V(consts.LogLevelWarning).Info("Cannot enable Validation state as podSelector is empty")
+		return m
+	}
+	m.ValidationManager = NewValidationManager(m.K8sInterface, m.Log, m.EventRecorder, m.NodeUpgradeStateProvider,
+		podSelector)
+	m.validationStateEnabled = true
+	return m
+}
+
+// IsPodDeletionEnabled returns true if 'pod-deletion' state is enabled
+func (m *CommonUpgradeManagerImpl) IsPodDeletionEnabled() bool {
+	return m.podDeletionStateEnabled
+}
+
+// IsValidationEnabled returns true if 'validation' state is enabled
+func (m *CommonUpgradeManagerImpl) IsValidationEnabled() bool {
+	return m.validationStateEnabled
+}
+
+// GetCurrentUnavailableNodes returns all nodes that are not in ready state
+// TODO: Drop ctx as it's not used
```

Comment:
What's the plan for dropping `ctx`? Can we deprecate this method and add another one without context?
https://go.dev/wiki/Deprecated